src/Control/Monad/Fail/Hoist.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE CPP                    #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs                  #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+
+-- | 'HoistFail extends 'MonadFail with 'hoistFail, which enables lifting
+-- of partiality types such as 'Maybe' and @'Either' e@ into the monad.
+module Control.Monad.Fail.Hoist
+  ( HoistFail(..)
+  , hoistFail_
+  , hoistFailE
+  , hoistFailE_
+  , (<%!>)
+  , (<!>)
+  ) where
+
+import           Prelude hiding (fail)
+
+import           Control.Monad.Fail         (MonadFail (..))
+
+import           Data.Either                (Either, either)
+
+class Monad m => HoistFail t m e | t -> e where
+
+  -- | Given a conversion from the error in @t a@ to @String@, we can hoist the
+  -- computation into @m@.
+  --
+  -- @
+  -- 'hoistFail' :: 'MonadFail' m => (() -> e) -> 'Maybe'       a -> m a
+  -- 'hoistFail' :: 'MonadFail' m => (a  -> e) -> 'Either'  a   b -> m b
+  -- @
+  hoistFail
+    :: (e -> String)
+    -> t a
+    -> m a
+
+instance MonadFail m => HoistFail Maybe m () where
+  hoistFail f = maybe (fail $ f ()) pure
+
+instance MonadFail m => HoistFail (Either e) m e where
+  hoistFail f = either (fail . f) pure
+
+hoistFail_ :: HoistFail t m String => t a -> m a
+hoistFail_ = hoistFail id
+
+hoistFailE :: MonadFail m => (e -> String) -> Either e a -> m a
+hoistFailE = hoistFail
+
+hoistFailE_ :: MonadFail m => Either String a -> m a
+hoistFailE_ = hoistFail_
+
+-- | A flipped synonym for 'hoistFail'.
+--
+-- @
+-- ('<%!>') :: 'MonadFail' m => 'Maybe'       a -> (() -> e) ->           m a
+-- ('<%!>') :: 'MonadFail' m => 'Either'  a   b -> (a  -> e) ->           m b
+-- @
+(<%!>)
```

Comment:
Bikeshed: I am not thrilled by these operator names, but I cannot think of better ones. `(<!>)` is the semigroupoids version of `(<|>)`, and the `!` is already used in `Control.Monad.Error.Hoist` to mean "with monadic effect".

`#` kinda looks like a bad 'F', and there [aren't very many occurrences of `(<#>)`](https://hoogle.haskell.org/?hoogle=%3C%23%3E) to mess with an existing mnemonic. WDYT about `(<%#>)`, `(<%!#>)`, `(<#>)`, and `<#!>`)?

Reply:
Sure, I'll change it.